go/pkg/cache/singleflightcache/singleflightcache.go

```diff
@@ -5,86 +5,42 @@
 package singleflightcache
 
 import (
-	"fmt"
 	"sync"
 )
 
 // Cache is a cache that supports single-flight value computation.
 type Cache struct {
-	// store is the actual cache where data is stored.
 	store sync.Map
-	// locks is a map of locks per key. This map is used to ensure only one reader/writer of
-	// the key can have write access.
-	locks sync.Map
-	// mu is a mutex to ensure exclusive access to the locks map in the case of Delete.
-	mu sync.RWMutex
+}
+
+type entry struct {
+	compute sync.Once
```

Comment:
I wonder if sync.Once is actually better than RWMutex. In the [implementation](https://golang.org/src/sync/once.go), seems like if there are multiple concurrent readers trying to access the same cache key (which is common in the execution case as the first batch of actions attempt to digest common toolchain files), readers might all enter `doSlow()`, and serialize on a lock of the internal mutex. Whereas with an [RWMutex](https://golang.org/src/sync/rwmutex.go), once the writer finishes writing, a semaphore is released with the reader count, unblocking all readers at once.

Reply:
I never compared their performance, but `sync.Once` was created exactly for this use case (ensure that the func runs only once). `RWMutex` is a more general thing, i.e. supports more use cases, so (hand wavy) it _should_ be slower.

Reply:
more concretely, there is only one place here where we lock, so we'd have to do `Lock()` (as opposed to `RLock()`), which is the same thing as not using `RWMutex`.